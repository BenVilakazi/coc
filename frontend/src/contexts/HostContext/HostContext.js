import React, { createContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import useReducerMiddleware from '../useReducerMiddleware';
import HostReducer from './HostReducer';
import socketInstance from '../../socket/socket';
import hostReducerMiddleware from './hostReducerMiddleware';
import config from '../../config';
import AlertModal from '../../components/Modal/AlertModal';

const propTypes = {
  children: PropTypes.node.isRequired,
};

const { emitter } = socketInstance;

const initialState = {
  gameState: 'waiting-for-lobby',
  lobbyID: '',
  players: {},
  playerIDs: [],
  gameSettings: {
    maxPlayers: config.maxPlayers.default,
    winningScore: config.winningScore.default,
    selectedPacks: config.initialSelectedPack,
    handSize: config.handSize,
    winnerScreenDisplayTime: config.winnerScreenDisplayTime,
    hideJoinCode: false,
  },
  deck: { black: [], white: [] },
  loading: [],
  czarSelection: '',
};

export const HostContext = createContext();

function HostProvider({ children }) {
  const [state, dispatch] = useReducerMiddleware(
    hostReducerMiddleware,
    HostReducer,
    initialState,
  );

  const [socketHasError, setSocketHasError] = useState(false);

  function createMessageHandler(currentState) {
    const czarId =
      currentState.playerIDs.length > 0 &&
      currentState.playerIDs.find(
        (playerId) => currentState.players[playerId].isCzar,
      );

    return ({ event, payload, sender }) => {
      switch (event) {
        case 'player-connected': {
          const allPlayers = [...Object.values(currentState.players)];
          const existingPlayer = allPlayers.find(
            (player) =>
              player.name.toUpperCase() === payload.playerName.toUpperCase(),
          );

          if (!existingPlayer) {
            return dispatch({ type: 'PLAYER_CONNECTED', payload });
          }

          if (existingPlayer.status === 'disconnected') {
            return dispatch({ type: 'PLAYER_RECONNECTED', payload });
          }

          return dispatch({ type: 'SEND_NAME_TAKEN_MESSAGE', payload });
        }
        case 'player-disconnected':
          return dispatch({
            type: 'PLAYER_DISCONNECTED',
            payload: {
              ...payload,
              czarId,
            },
          });

        case 'preview-winner':
          return dispatch({ type: 'PREVIEW_WINNER', payload });

        case 'winner-selected':
          return dispatch({ type: 'WINNER_SELECTED', payload });

        case 'lobby-created':
          return dispatch({ type: 'SET_LOBBY_ID', payload });

        case 'black-card-selected':
          return dispatch({ type: 'SELECT_BLACK_CARD', payload });

        case 'join-code-shuffled':
          return dispatch({ type: 'UPDATE_JOIN_CODE', payload });

        case 'player-submit':
          return dispatch({
            type: 'PLAYER_SUBMIT',
            payload: { ...payload, playerId: sender },
          });

        case 'socket-connection-error':
          dispatch({ type: 'UPDATE_JOIN_CODE', payload: { lobbyID: 'ERROR' } });
          return setSocketHasError(true);

        default:
          return undefined;
      }
    };
  }

  useEffect(() => {
    const handleMessage = createMessageHandler(state);
    emitter.on('message', handleMessage);
    return () => {
      emitter.off('message', handleMessage);
    };
  }, [state]);

  useEffect(() => {
    const stagedPlayerIDs = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 'staging',
    );

    const reconnectingPlayersIds = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 're-connected',
    );

    if (reconnectingPlayersIds.length > 0) {
      const reConnectingPlayers = reconnectingPlayersIds.reduce(
        (acc, id) => ({ ...acc, [id]: state.players[id] }),
        {},
      );

      dispatch({
        type: 'SEND_CARDS_TO_PLAYERS',
        payload: {
          selectedBlackCard: state.selectedBlackCard,
          players: reConnectingPlayers,
          playerIDs: reconnectingPlayersIds,
        },
      });
    }

    if (stagedPlayerIDs.length) {
      const numberOfPlayersExceedingLimit =
        state.playerIDs.length - state.gameSettings.maxPlayers;

      if (numberOfPlayersExceedingLimit > 0) {
        const playerIDsExceedingLimit = state.playerIDs.slice(
          -numberOfPlayersExceedingLimit,
        );

        dispatch({
          type: 'TOO_MANY_PLAYERS',
          payload: {
            players: playerIDsExceedingLimit,
          },
        });
      } else {
        const message =
          state.gameState === 'waiting-for-players'
            ? {
                big: "You've joined the lobby",
                small: 'Please wait for the host to start the game',
              }
            : {
                big: 'A round is in progress',
                small: 'You will join the next round automatically',
              };

        dispatch({
          type: 'SEND_PLAYER_CONNECTED_MESSAGES',
          payload: {
            players: stagedPlayerIDs,
            message,
          },
        });
      }
    }
  }, [state.playerIDs]);

  useEffect(() => {
    // if there is one player AND they are not waiting for the game to begin, end the game
    if (
      state.playerIDs.filter(
        (playerID) => state.players[playerID].status === 'playing',
      ).length === 1 &&
      state.gameState !== 'game-over'
    ) {
      dispatch({
        type: 'GAME_OVER',
        payload: { gameWinner: state.playerIDs[0], playerIDs: state.playerIDs },
      });
    }
  }, [state.playerIDs, state.players]);

  return (
    <HostContext.Provider value={{ state, dispatch }}>
      {children}
      {socketHasError && (
        <AlertModal
          bigText="SOCKET ERROR"
          smallText="Please try again later"
          buttonText="Click anywhere to restart"
          onClick={() => window.location.reload()}
        />
      )}
    </HostContext.Provider>
  );
}

HostProvider.propTypes = propTypes;

export default HostProvider;
