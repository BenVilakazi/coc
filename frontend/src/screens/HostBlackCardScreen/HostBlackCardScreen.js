import React, { useContext, useEffect } from 'react';
import styled from 'styled-components';
import HostLayout from '../../layouts/HostLayout';
import { HostContext } from '../../contexts/HostContext/HostContext';
import PlayerList from '../../components/PlayerList/PlayerList';
import InGameSettingsModal from '../../components/HostSettingsMenu/InGameSettingsModal.js';
import JoinCode from '../../components/JoinCode/JoinCode';
import BlackCard from '../../components/Cards/BlackCard';

const LeftPanelWrapper = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;

  .player-list-wrapper {
    overflow-x: auto;
    flex-grow: 1;
  }

  .buttons-wrapper {
    display: block;
    margin-bottom: 20px;
  }

  .bottom-left-wrapper {
    margin: 20px auto 32px;
    position: relative;
    bottom: 0;
    left: 0;
    right: 0;
  }

  .join-code-wrapper {
    display: flex;
    justify-content: center;
  }

  .host-pregame-button {
    padding: 0;
    background-color: var(--primary-background-color);
    color: var(--primary-text-color);
    margin: auto auto 12px;
  }
`;

const RightPanelWrapper = styled.div`
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;

  .card-display {
    height: 100%;

    margin-top: 64px;

    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .czar-display {
    font-size: 24px;
    font-weight: 600;

    line-height: 24px;

    .czar-name {
      margin-bottom: 48px;
      padding-left: 16px;

      font-size: 36px;
      font-weight: 600;
    }
  }
`;

function LeftPanel() {
  const { state } = useContext(HostContext);
  const { lobbyID } = state;

  return (
    <LeftPanelWrapper>
      <div className="player-list-wrapper">
        <PlayerList />
      </div>
      <div className="bottom-left-wrapper">
        <div className="join-code-wrapper">
          <JoinCode
            loading={state.loading.includes('join-code')}
            code={lobbyID}
          />
        </div>
      </div>
    </LeftPanelWrapper>
  );
}

function RightPanel() {
  const { state } = useContext(HostContext);

  const { selectedBlackCard, players, playerIDs } = state;

  const czar = players[playerIDs.find((player) => players[player].isCzar)];

  return (
    <RightPanelWrapper>
      <div className="card-display">
        <div className="czar-display">
          <p>CZAR:</p>
          <p data-testid="czar-name" className="czar-name">
            {czar?.name.toUpperCase()}
          </p>
        </div>

        <BlackCard pickCount={selectedBlackCard.pick} data-test-id="black-card">
          {selectedBlackCard.text}
        </BlackCard>
      </div>
    </RightPanelWrapper>
  );
}

function HostBlackCardScreen() {
  const { state, dispatch } = useContext(HostContext);

  const { players, playerIDs, selectedBlackCard, gameState } = state;
  const czar = playerIDs.find((id) => players[id].isCzar);

  useEffect(async () => {
    if (gameState === 'waiting-to-receive-cards') {
      await dispatch({
        type: 'SEND_CARDS_TO_PLAYERS',
        payload: { players, playerIDs, selectedBlackCard },
      });
      if (czar) {
        await dispatch({
          type: 'NOTIFY_CZAR',
          payload: {
            players,
            playerIDs,
          },
        });
      }
    }
  }, [state.gameState, czar]);

  useEffect(async () => {
    // game state will change when players have all submitted cards
    if (gameState === 'czar-select-winner') {
      await dispatch({
        type: 'CZAR_SELECT_WINNER',
        payload: {
          players,
          playerIDs,
        },
      });
    }
  }, [gameState]);

  return (
    <HostLayout
      className="primary-background"
      left={<LeftPanel />}
      right={<RightPanel />}
      modal={<InGameSettingsModal />}
    />
  );
}

export default HostBlackCardScreen;
