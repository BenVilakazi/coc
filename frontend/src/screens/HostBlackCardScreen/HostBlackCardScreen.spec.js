import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { HostContext } from '../../contexts/HostContext/HostContext';
import HostBlackCardScreen from './HostBlackCardScreen';

jest.mock('../../components/GameSettings/GameSettings', () => () => (
  <div data-testid="game-settings" />
));

describe('Host Black Card Screen', () => {
  describe('rendering', () => {
    const dispatch = jest.fn();
    const state = {
      gameState: 'waiting-to-receive-cards',
      lobbyID: 'ABCD',
      players: {
        ID1: {
          name: 'foo',
          score: 0,
          isCzar: false,
          submittedCards: [0, 1],
          cards: [
            { text: 'aaaa' },
            { text: 'bbbb' },
            { text: 'cccc' },
            { text: 'dddd' },
          ],
        },
        ID2: {
          name: 'bar',
          score: 0,
          isCzar: true,
          submittedCards: [2],
          cards: [],
        },
        ID3: {
          name: 'baz',
          score: 0,
          isCzar: false,
          submittedCards: [1, 2],
          cards: [
            { text: 'eeee' },
            { text: 'ffff' },
            { text: 'gggg' },
            { text: 'hhhh' },
          ],
        },
      },
      selectedBlackCard: {
        text: 'Test Black Card',
        pick: 1,
        pack: 0,
      },
      playerIDs: ['ID1', 'ID2', 'ID3'],
      gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
      loading: [],
    };

    it('renders', () => {
      const tree = renderer
        .create(
          <HostContext.Provider value={{ state, dispatch }}>
            <HostBlackCardScreen />
          </HostContext.Provider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders the black card', () => {
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostBlackCardScreen />
        </HostContext.Provider>,
      );

      expect(screen.queryByTestId('black-card')).toBeInTheDocument();
    });

    it("renders the czar's name in the czar display", () => {
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostBlackCardScreen />
        </HostContext.Provider>,
      );

      expect(screen.queryByTestId('czar-name')).toBeInTheDocument();
      expect(screen.queryByTestId('czar-name')).toHaveTextContent('BAR');
    });

    it("does not render the loading indicator when loading state does not contain 'join-code'", () => {
      render(
        <HostContext.Provider
          value={{
            state: { ...state, loading: ['aaa', 'test', 'join-cod'] },
            dispatch,
          }}
        >
          <HostBlackCardScreen />
        </HostContext.Provider>,
      );

      expect(screen.queryByTestId('loader')).not.toBeInTheDocument();
    });

    it("renders the loading indicator when loading state contains 'join-code'", () => {
      render(
        <HostContext.Provider
          value={{ state: { ...state, loading: ['join-code'] }, dispatch }}
        >
          <HostBlackCardScreen />
        </HostContext.Provider>,
      );

      expect(screen.getByTestId('loader')).toBeInTheDocument();
    });
  });

  describe('dispatch', () => {
    it('sends the cards to players when the game state is waiting-to-receive-cards', () => {
      const dispatch = jest.fn();
      const state = {
        gameState: 'waiting-to-receive-cards',
        lobbyID: 'ABCD',
        players: {
          ID1: {
            name: 'foo',
            score: 0,
            isCzar: false,
            submittedCards: [0, 1],
            cards: [
              { text: 'aaaa' },
              { text: 'bbbb' },
              { text: 'cccc' },
              { text: 'dddd' },
            ],
          },
          ID2: {
            name: 'bar',
            score: 0,
            isCzar: true,
            submittedCards: [2],
            cards: [],
          },
          ID3: {
            name: 'baz',
            score: 0,
            isCzar: false,
            submittedCards: [1, 2],
            cards: [
              { text: 'eeee' },
              { text: 'ffff' },
              { text: 'gggg' },
              { text: 'hhhh' },
            ],
          },
        },
        selectedBlackCard: {
          text: 'Test Black Card',
          pick: 1,
          pack: 0,
        },
        playerIDs: ['ID1', 'ID2', 'ID3'],
        gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
        loading: [],
      };
      const { players, playerIDs, selectedBlackCard } = state;

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostBlackCardScreen />
        </HostContext.Provider>,
      );

      expect(dispatch).toHaveBeenCalledWith({
        type: 'SEND_CARDS_TO_PLAYERS',
        payload: { players, playerIDs, selectedBlackCard },
      });
    });

    it('notifies the czar when the game state is waiting-to-receive-cards', async () => {
      const dispatch = jest.fn();
      const state = {
        gameState: 'waiting-to-receive-cards',
        lobbyID: 'ABCD',
        players: {
          ID1: {
            name: 'foo',
            score: 0,
            isCzar: false,
            submittedCards: [0, 1],
            cards: [
              { text: 'aaaa' },
              { text: 'bbbb' },
              { text: 'cccc' },
              { text: 'dddd' },
            ],
          },
          ID2: {
            name: 'bar',
            score: 0,
            isCzar: true,
            submittedCards: [2],
            cards: [],
          },
          ID3: {
            name: 'baz',
            score: 0,
            isCzar: false,
            submittedCards: [1, 2],
            cards: [
              { text: 'eeee' },
              { text: 'ffff' },
              { text: 'gggg' },
              { text: 'hhhh' },
            ],
          },
        },
        selectedBlackCard: {
          text: 'Test Black Card',
          pick: 1,
          pack: 0,
        },
        playerIDs: ['ID1', 'ID2', 'ID3'],
        gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
        loading: [],
      };
      const { players, playerIDs } = state;

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostBlackCardScreen />
        </HostContext.Provider>,
      );

      await waitFor(() => {
        expect(dispatch).toHaveBeenCalledTimes(2);
      });

      expect(dispatch).toHaveBeenNthCalledWith(2, {
        type: 'NOTIFY_CZAR',
        payload: { players, playerIDs },
      });
    });

    it('dispatches the CZAR_SELECT_WINNER action when the game state is set to czar-select-winner', () => {
      const dispatch = jest.fn();
      const state = {
        gameState: 'czar-select-winner',
        lobbyID: 'ABCD',
        players: {
          ID1: {
            name: 'foo',
            score: 0,
            isCzar: false,
            submittedCards: [0, 1],
            cards: [
              { text: 'aaaa' },
              { text: 'bbbb' },
              { text: 'cccc' },
              { text: 'dddd' },
            ],
          },
          ID2: {
            name: 'bar',
            score: 0,
            isCzar: true,
            submittedCards: [2],
            cards: [],
          },
          ID3: {
            name: 'baz',
            score: 0,
            isCzar: false,
            submittedCards: [1, 2],
            cards: [
              { text: 'eeee' },
              { text: 'ffff' },
              { text: 'gggg' },
              { text: 'hhhh' },
            ],
          },
        },
        selectedBlackCard: {
          text: 'Test Black Card',
          pick: 1,
          pack: 0,
        },
        playerIDs: ['ID1', 'ID2', 'ID3'],
        gameSettings: { maxPlayers: 8, winningScore: 7, selectedPacks: [] },
        loading: [],
      };
      const { players, playerIDs } = state;

      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostBlackCardScreen />
        </HostContext.Provider>,
      );

      expect(dispatch).toHaveBeenCalledWith({
        type: 'CZAR_SELECT_WINNER',
        payload: { players, playerIDs },
      });
    });
  });
});
