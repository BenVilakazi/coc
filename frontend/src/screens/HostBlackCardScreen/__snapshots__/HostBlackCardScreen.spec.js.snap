// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Host Black Card Screen rendering renders 1`] = `
.c1 {
  background-image: url(inactiveHamburger.svg);
  background-size: cover;
  background-color: transparent;
  border: 0;
  width: 25px;
  height: 25px;
  cursor: pointer;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .hamburger-container {
  position: absolute;
  right: 17px;
  top: 11px;
  z-index: 10100;
}

.c0 .host-layout-header {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  height: 50px;
}

.c0 .host-layout-header h2 {
  position: relative;
  bottom: -1px;
  font-size: 1rem;
  line-height: 0.8rem;
}

.c0 .components {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

.c0 .components .left {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 200px;
  background-color: var(--primary-background-color);
  box-shadow: 0px 4px 12px rgba(0,0,0,0.25);
}

.c0 .components .right {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  position: relative;
  min-width: 440px;
}

.c2 {
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 486px;
  padding: 40px;
  background-color: var(--primary-background-color);
  position: relative;
}

.c4 > div:not(:last-child) {
  margin-bottom: 24px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c5 .player-info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 24px 0 26px;
  padding: 7px 0;
  border-bottom: 1px solid #ccc;
  border-radius: 0;
  background-color: var(--primary-background-color);
}

.c5 .player-info h1 {
  text-transform: uppercase;
  font-size: 24px;
  font-weight: 900;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1.5em;
  color: var(--primary-text-color);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5 .player-info span {
  color: var(--primary-text-color);
  font-size: 24px;
  font-weight: 900;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c7 .player-info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 12px 0 26px;
  padding: 7px 12px;
  border-bottom: 0;
  border-radius: 4px;
  background-color: var(--secondary-background-color);
}

.c7 .player-info h1 {
  text-transform: uppercase;
  font-size: 24px;
  font-weight: 900;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1.5em;
  color: var(--secondary-text-color);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7 .player-info span {
  color: var(--secondary-text-color);
  font-size: 24px;
  font-weight: 900;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  line-height: 29px;
}

.c8 .join-code-title {
  font-size: 18px;
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  font-weight: 700;
  width: 97px;
  height: 21px;
}

.c8 .join-code {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 56px;
  font-weight: bold;
  color: var(--secondary-text-color);
  background-color: var(--secondary-background-color);
  width: 305px;
  height: 70px;
}

.c10 {
  position: relative;
  width: 390px;
  height: 540px;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
  border-radius: 18px;
  font-weight: 700;
  font-size: 36px;
  padding: 45px 21px 0 21px;
  box-shadow: 4px 4px 24px rgba(0,0,0,0.3);
}

.c10 .pickText {
  font-size: 24px;
  margin-right: 7px;
}

.c10 .layeredCards {
  position: absolute;
  bottom: 18px;
  left: 18px;
}

.c10 .pickCount {
  position: absolute;
  bottom: 8px;
  right: 21px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: var(--primary-text-color);
  background: var(--primary-background-color);
  width: 20px;
  height: 20px;
  border-radius: 50%;
  font-weight: bold;
  font-size: 14px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.c3 .player-list-wrapper {
  overflow-x: auto;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 .buttons-wrapper {
  display: block;
  margin-bottom: 20px;
}

.c3 .bottom-left-wrapper {
  margin: 20px auto 32px;
  position: relative;
  bottom: 0;
  left: 0;
  right: 0;
}

.c3 .join-code-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 .host-pregame-button {
  padding: 0;
  background-color: var(--primary-background-color);
  color: var(--primary-text-color);
  margin: auto auto 12px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
}

.c9 .card-display {
  height: 100%;
  margin-top: 64px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 .czar-display {
  font-size: 24px;
  font-weight: 600;
  line-height: 24px;
}

.c9 .czar-display .czar-name {
  margin-bottom: 48px;
  padding-left: 16px;
  font-size: 36px;
  font-weight: 600;
}

@media (min-width:680px) {
  .c0 .host-layout-header {
    height: 130px;
  }

  .c0 .components .left {
    min-width: 420px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 2.5rem;
    margin-bottom: -4px;
    line-height: 2rem;
  }
}

@media (min-width:1600px) {
  .c0 .components .left {
    min-width: 700px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 3.5rem;
    margin-bottom: -35px;
    line-height: 7rem;
  }
}

@media (min-width:3200px) {
  .c0 .host-layout-header {
    height: 200px;
  }

  .c0 .components .left {
    min-width: 1100px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 6rem;
    margin-bottom: -25px;
    line-height: 7rem;
  }
}

<div
  className="c0 primary-background"
>
  <div
    className="hamburger-container"
  >
    <button
      className="c1"
      onClick={[Function]}
    />
  </div>
  <div
    className="components"
  >
    <div
      className="left"
    >
      <header
        className="c2 host-layout-header"
      >
        <h2>
          CARDS OF CAROUSAL
        </h2>
      </header>
      <div
        className="c3"
      >
        <div
          className="player-list-wrapper"
        >
          <div
            className="c4"
            data-testid="playerList-container"
          >
            <div
              className="c5"
              data-testid="row-foo"
            >
              <img
                alt="card icon"
                data-testid="icon-foo"
                src="offset-white-card-stack-icon.svg"
                style={
                  Object {
                    "visibility": "visible",
                  }
                }
              />
              <div
                className="player-info"
              >
                <h1>
                  foo
                </h1>
                <div
                  className="c6"
                >
                  <svg
                    data-testid="tally-svg"
                    fill="none"
                    height="30"
                    viewBox="0 0 42 30"
                    width="42"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g />
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="c7"
              data-testid="row-bar"
            >
              <img
                alt="card icon"
                data-testid="icon-bar"
                src="black-card-icon.svg"
                style={
                  Object {
                    "visibility": "visible",
                  }
                }
              />
              <div
                className="player-info"
              >
                <h1>
                  bar
                </h1>
                <div
                  className="c6"
                >
                  <svg
                    data-testid="tally-svg"
                    fill="none"
                    height="30"
                    viewBox="0 0 42 30"
                    width="42"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g />
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="c5"
              data-testid="row-baz"
            >
              <img
                alt="card icon"
                data-testid="icon-baz"
                src="offset-white-card-stack-icon.svg"
                style={
                  Object {
                    "visibility": "visible",
                  }
                }
              />
              <div
                className="player-info"
              >
                <h1>
                  baz
                </h1>
                <div
                  className="c6"
                >
                  <svg
                    data-testid="tally-svg"
                    fill="none"
                    height="30"
                    viewBox="0 0 42 30"
                    width="42"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="bottom-left-wrapper"
        >
          <div
            className="join-code-wrapper"
          >
            <div
              className="c8"
            >
              <p
                className="join-code-title"
              >
                JOIN CODE:
              </p>
              <div
                className="join-code"
                data-testid="join-code"
              >
                <p>
                  ABCD
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="right"
    >
      <div
        className="c9"
      >
        <div
          className="card-display"
        >
          <div
            className="czar-display"
          >
            <p>
              CZAR:
            </p>
            <p
              className="czar-name"
              data-testid="czar-name"
            >
              BAR
            </p>
          </div>
          <div
            className="c10"
            data-testid="black-card"
          >
            <span>
              Test Black Card
            </span>
            <div
              className="layeredCards"
            >
              <div
                style={
                  Object {
                    "position": "relative",
                  }
                }
              >
                <img
                  alt="offset black and white card stack"
                  data-testid="layerCard"
                  src="cards-stack-offset-dark-icon.svg"
                />
              </div>
            </div>
            <div
              className="pickCount"
            >
              <span
                className="pickText"
              >
                PICK
              </span>
              <div
                className="c11"
              >
                1
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
